tools/oeedger8r/Emitter.ml

```diff
@@ -483,435 +524,371 @@ let get_cast_from_mem_expr (ptype, decl) =
+    ^ fd.fname ^ "("
+  ; String.concat ",\n    "
+      (List.map
+         (fun (ptype, decl) ->
+           let cast_expr = get_cast_from_mem_expr (ptype, decl) in
+           sprintf "    %spargs_in->%s" cast_expr decl.identifier )
+         fd.plist)
+    ^ ");" ]
 
-let oe_gen_copy_outputs (os : out_channel) (fd : func_decl) =
-  let gen_free_buffer (ptype, decl) =
-    match ptype with
-    | PTPtr (atype, ptr_attr) ->
-        if ptr_attr.pa_chkptr then
-          match ptr_attr.pa_direction with
-          | PtrOut | PtrInOut ->
-              fprintf os "    if (args.%s)\n" decl.identifier ;
-              fprintf os "        memcpy(args.%s, enc_args.%s, %s);\n"
-                decl.identifier decl.identifier
-                (oe_get_param_size (ptype, decl, "args."))
-          | _ -> ()
-        else ()
-    | _ -> ()
-    (* Non pointer arguments *)
+let oe_gen_in_and_inout_setters (plist : pdecl list) =
+  let params =
+    List.map
+      (fun (ptype, decl) ->
+        let size = oe_get_param_size (ptype, decl, "pargs_in->") in
+        let tystr = get_cast_to_mem_expr (ptype, decl) false in
+        sprintf "OE_SET_%s_POINTER(%s, %s, %s);"
+          (if is_in_ptr (ptype, decl) then "IN" else "IN_OUT")
+          decl.identifier size tystr )
+      (List.filter (fun p -> is_in_ptr p || is_inout_ptr p) plist)
   in
-  fprintf os "\n    /* Copy output buffers */\n" ;
-  List.iter gen_free_buffer fd.plist ;
-  fprintf os "\n"
+  "    "
+  ^ String.concat "\n    "
+      [ "/* Set in and in-out pointers. */"
+      ; ( if params <> [] then String.concat "\n    " params
+        else "/* There were no in nor in-out parameters. */" ) ]
 
-let oe_gen_call_function (os : out_channel) (fd : func_decl) =
+let oe_gen_out_and_inout_setters (plist : pdecl list) =
   let params =
     List.map
-      (fun (pt, decl) ->
-        sprintf "%spargs_in->%s"
-          (get_cast_from_mem_expr (pt, decl))
-          decl.identifier )
-      fd.plist
-  in
-  let params_str = "(\n        " ^ String.concat ",\n        " params ^ ")" in
-  let ret_str =
-    match fd.rtype with Void -> "" | _ -> "pargs_out->_retval = "
+      (fun (ptype, decl) ->
+        let size = oe_get_param_size (ptype, decl, "pargs_in->") in
+        let tystr = get_cast_to_mem_expr (ptype, decl) false in
+        sprintf "OE_%s_POINTER(%s, %s, %s);"
+          ( if is_out_ptr (ptype, decl) then "SET_OUT"
+          else "COPY_AND_SET_IN_OUT" )
+          decl.identifier size tystr )
+      (List.filter (fun p -> is_out_ptr p || is_inout_ptr p) plist)
   in
-  let call_str = ret_str ^ fd.fname ^ params_str in
-  fprintf os "    /* Call user function */\n" ;
-  fprintf os "    %s;\n" call_str
+  "    "
+  ^ String.concat "\n    "
+      [ "/* Set out and in-out pointers. */"
+      ; "/* In-out parameters are copied to output buffer. */"
+      ; ( if params <> [] then String.concat "\n    " params
+        else "/* There were no out nor in-out parameters. */" ) ]
```

Comment:
Can you double check the grammar here? 
We were taught that it was "neither..nor"

Reply:
Yeah it [seems](https://www.wikihow.com/Use-nor) to [be](https://thewritepractice.com/how-to-use-either-neither-or-and-nor-correctly/) correct. It's not as common, but you can use 'nor' without 'neither' if it's paired with another negative (in this case, 'no').